### DISCUSSION TITLE
qiankun 3.0 Roadmap

### DISCUSSION THREAD

**Discussion by kuitos** [umijs/qiankun#1378] [Announcements]
### Core
  - [x] Standalone App Loader Module
    - [ ] ~API design (follows import-html-entry)~
    - [x] Streaming for best loading  performance
    - [ ] RegExp replaced with DOMParser
    - [x] Exception protection
  - [ ] Standalone Sandbox Module
    - [x] Pluggable, API refers to: https://github.com/tc39/proposal-realms
    - [ ] Default with scoped-style approach
      - [ ] Collaborate with https://github.com/samthor/scoped @chessl 
  - [x] [New logo](https://github.com/umijs/qiankun/discussions/1987)
  - [x] Support bundless (Vite)
  - [ ] Support Webpack 5 Module Federation
  - [x] Support CSP security strategy

### UI Library Bindings
  - [x] React
    - [x] MicroApp
    - [x] MicroAppLink 
  - [x] Vue
  - [ ] Angular

### Developer Friendly
  - [x] Webpack plugin
  - [x] UMD to window bundle
  - [x] Scaffold by cli
  - [ ] Upgrade guides

### Engineering
  - [x] Monorepo @ice-zjchen 
    - [x] Release note script
    - [x] ESLint/Prettier
  - [ ] Tests
    - [ ] UT
    - [ ] E2E
  - [ ] Benchmark https://github.com/Polymer/tachometer

### Extra Credits
  - [ ] Qiankun inspector for Chrome @chessl 

## Improvement
- [x] Vue performance issues
- [ ] Official docs
- [ ] Online examples

## Breaking Changes
- [x] Remove the deprecated `render` config
- [x] Deprecate ShadowDOM approach
- [x] Remove `globalState` API (provide compatible solution?)
- [x] Remove `addGlobalUncaughtErrorHandler` qiankun-irrelevant  APIs(or pluggable?)

**Comment by howel52** [comment]
> 插件化沙箱，支持自己添加沙箱逻辑

怎么理解插件化沙箱，提供一些官方的沙箱策略由业务插拔？ 用户自己该如何加沙箱逻辑，是暴露一些钩子出来么？

如果是以上这种理解的话，感觉可以设计下沙箱插件的 api，官方的沙箱插件们也都走这套标准，加上 sandbox 独立发包后可能就成沙箱方向的事实标准了

**Reply by kuitos** [reply]
官方提供默认的沙箱实现，用户可以选择增加自己的逻辑变得更 strict，也可以选择关掉一些内置实现变得更 loose。
具体如何实现现在其实没想好，我会单独开一个帖子

**Reply by codcer** [reply]
沙箱插件化,  是以lifecycle, 注入对应的钩子api,,,还是完全插拔?

**Comment by YuYang019** [comment]
shadowDOM 为啥不推荐了呢

**Reply by kuitos** [reply]
shadowdom 对其他依赖库要求比较高，比如 react 需要 17+，各种 UI 库也需要适配支持，整体的使用成本远高于目前的 experimentalStyleIsolation 方案。3.0 是计划将 experimentalStyleIsolation 方案作为标准样式隔离方案推荐使用，shadowdom 变成 legacy 策略

**Reply by Calvin-for-AI** [reply]
shadowDOM可以一直作为保留策略吧，随着时间推移逐步会成为事实标准的

**Reply by JingweiXiong** [reply]
experimentalStyleIsolation 方案也会有局限性？比如部分样式是应用到全局的？@kuitos

**Comment by youpengcheng** [comment]
请问3.0什么时候能上

**Reply by ralfsqual** [reply]
同问，期待支持vite

**Reply by renwofei423** [reply]
同问，大概什么时候能出来？

**Reply by Jam1zhu** [reply]
同问。。

**Reply by wutongqianyu** [reply]
同上

**Reply by kang-Z-A** [reply]
+1

**Reply by 1514100951** [reply]
+1

**Reply by jiangdotgs** [reply]
++1

**Comment by ChanFerris** [comment]
qiankun3.0中会支持开启sandbox的module federation吗？

**Comment by canyuegongzi** [comment]
想咨询下3.0有具体的排期吗？

**Comment by sky124380729** [comment]
很期待能够支持vite，另外不是很理解"webpack5 module federation 支持"这句话，qiankun本身就是基座的思想，模块联邦是去中心化的思想，属于两种不同的设计模式才对吧。
我觉得在未来，新项目都用webpack5构建，直接使用module federation构建微前端，应该也用不到qiankun了吧。。。
我觉得乾坤最大的优势还是对于一些老项目(jsp)的增量升级友好，继续维护之前的jsp太恶心了。

**Reply by NexxLuo** [reply]
同问，qiankun和module federation相比，其优势或者说不同点在哪？

**Comment by vickmeng** [comment]
为什考虑移除globalState？有准确原因吗？

**Reply by kuitos** [reply]
globalState 不是合理的微前端通信方案，会加剧应用之间的耦合

**Reply by zhangvj** [reply]
那新的通信方案是什么？目前在使用的2.*版本已经在提示globalState 将会弃用了

**Reply by rubickecho** [reply]
同样关注，新的通信方案

**Reply by Mister-Duan** [reply]
+1

**Reply by uabidotfun** [reply]
+1

**Reply by Jerome-1010** [reply]
+1
V3是否会提供一个替代globalState的通信方法？

**Reply by KuiShang** [reply]
+1

**Reply by BTBMan** [reply]
+1

**Reply by Aeero** [reply]
直接主应用window上挂一个全局变量，类似微信小程序的wx

**Comment by humorHan** [comment]
有大致的时间规划吗

**Reply by jspao** [reply]
+1

**Comment by qiaozhu** [comment]
进度如何了？

**Reply by renwofei423** [reply]
同问。

**Comment by bobfw** [comment]
期待vite 中使用

**Comment by Dashuai-Black** [comment]
qiankun 有讨论群么

**Reply by longshao5520** [reply]
看官网，有钉钉群

**Comment by eightHundreds** [comment]
希望changelog不要用表情,看不懂. 直接用Feature, Fix....

**Comment by GitY-elf** [comment]
期待可以适配vite

**Comment by ChengranYan** [comment]
Qiankun inspector for Chrome   是不是可以加快些速度

**Comment by ac51300696** [comment]
想了解下vue性能优化那一项的具体内容，现在在vue2.0中的性能问题比较严重

**Comment by liuye1296** [comment]
预加载 在**配置较低的电脑**上面存在性能问题。 虽然使用到了 requestIdleCallback   但是预加载的子应用很多 **eval 执行太久** 会导致浏览器卡住   通过loader 函数 显示加载 都没法显示 被eval  占用太多时间。目前我们公司的是关闭预加载的。 自己在Worker 里面预先请求一遍所有资源 后续  切换子应用 虽然还会发送请求 但是会使用浏览器缓存  也达到了目的。 3.0 是否考虑 这种模式?

**Reply by kuitos** [reply]
预加载不会 eval 子应用，只会发请求拉资源

**Reply by liuye1296** [reply]
> 预加载不会 eval 子应用，只会发请求拉资源

考虑 用 Worker  去实现 预加载嘛?  我看nextjs 都是用Worker预加载资源的。 我现在自己手动实现的  希望V3 官方支持

**Reply by gwuhaolin** [reply]
> 预加载不会 eval 子应用，只会发请求拉资源

qiankun2的prefetch会拉资源 + 缓存JS内容、CSS内容、HTML内容到内存中，https://github.com/kuitos/import-html-entry/blob/cb92c1fb4990eb80473c89e35739fa985359c9c6/src/index.js#L84
子应用太多了会导致OOM，我们站遇到过这种线上问题。

我觉得prefetch可以分成两种策略：
- 只拉资源 
- 拉资源  + 缓存内容到内存中

**Comment by 316374157** [comment]
这个有没有预计的日期啊，好多对号都没勾选啊

**Comment by wangrongding** [comment]
蹲一个

**Comment by CasoMemory** [comment]
今天刷到了，mark 下，等待后续进展。

**Comment by Go-loneliness-of-man** [comment]
今天组内准备新模块上 vue3 + vite，查了一早上到这了，希望官方尽快更新

**Comment by qw789** [comment]
哈哈哈，就完成了新的logo

**Comment by aehyok** [comment]
蹲一下，看看什么时候能有动静

**Comment by pengkobe** [comment]
好吧，我也来个催更

**Comment by jackbi** [comment]
啥时候能更新呢🤔

**Comment by aehyok** [comment]
不出意外，今年应该没戏

**Comment by qiaozhu** [comment]
ε=(´ο｀*)))唉
不用等了，1年多了，就完成了一个新logo，感觉像是凉了。
目前qiankun也不是不能用，基于qiankun采用vue3+vite+vite-plugin-qiankun，我们自己团队有落地了一套方案。

其他方案：
方案1：基于webpack5 模块联邦   可以用emp框架
方案2：如果想要使用vite，那么使用这个vite版的模块联邦插件vite-plugin-federation，也是很不错的方案（推荐）

**Reply by rubickecho** [reply]
目前我们团队就是基于 webpack5 做的微前端方案

**Reply by carvinlo** [reply]
这边用的 piral 微前端方案，DDD 思想

**Reply by fantasticsoul** [reply]
也可以试试 [hel-micro](https://github.com/tnfe/hel) 方案

**Comment by reactjser** [comment]
进度怎么样了？

**Comment by jk4235** [comment]
any updates?

**Comment by zhengxif** [comment]
3.0快上了吗

**Comment by typistZxd** [comment]
听说v3沙箱用 ShadowRealm 来实现？

**Reply by ac51300696** [reply]
不可能的，用这个兼容直接爆炸

**Comment by yizhengfeng-jj** [comment]
”Standalone Sandbox Module“
沙盒独立发包这个有了么，是哪一个包哇

**Reply by carvinlo** [reply]
没有吧

**Comment by kuitos** [comment]
https://github.com/umijs/qiankun/pull/2473
提前预览版，欢迎围观😃

**Reply by carvinlo** [reply]
![3689F6B2](https://user-images.githubusercontent.com/5674761/234152070-80610d1d-5a45-4041-9aef-b7ef9d2e009a.png)
沙箱模块代码改动挺大，请问有哪方面改进么？沙箱模块可以单独拿来使用么？

**Reply by qiaozhu** [reply]
🤩  哇 终于！！！

**Reply by kuitos** [reply]
> ![3689F6B2](https://user-images.githubusercontent.com/5674761/234152070-80610d1d-5a45-4041-9aef-b7ef9d2e009a.png) 沙箱模块代码改动挺大，请问有哪方面改进么？沙箱模块可以单独拿来使用么？

是的，沙箱会更灵活定制型更强

**Reply by carvinlo** [reply]
太棒了，请问有计划将沙箱独立出一个库么？

**Reply by kuitos** [reply]
看上面的pr，沙箱就是独立的包

**Reply by carvinlo** [reply]
盼望预览版

**Reply by Vincent1993** [reply]
🐮 终于等到你

**Reply by gjmhandsome** [reply]
看Support bundless (Vite) 已经完成，可以提供相关案例吗？参考一下实现思路，目前用的qiankun-vite-plugin的方式

**Comment by wangronglei** [comment]
期待

**Comment by ifreeWorld** [comment]
期待

**Comment by gdnwr** [comment]
希望升级3.0不用对子项目做任何处理。现有生产环境是主项目一个，子项目几十个，而且子项目分不同的部门维护。

**Comment by fei-king** [comment]
啥时能用啊，这都两年了

**Comment by fei-king** [comment]
该邮箱已经转发至阿里去邮

**Comment by guopingxiao** [comment]
2023年国庆了，请问下，什么时候发布呀~~~~

**Comment by fei-king** [comment]
该邮箱已经转发至阿里去邮

**Comment by kuitos** [comment]
想参与 3.0 开发的小伙伴可以看这里 https://github.com/umijs/qiankun/discussions/2688 🤩
